crates/rustc_codegen_nvvm/src/context.rs

```diff
@@ -267,7 +271,31 @@ impl<'ll, 'tcx> CodegenCx<'ll, 'tcx> {
         }
 
         if !is_mutable && self.type_is_freeze(ty) {
-            AddressSpace(4)
+            if !self.codegen_args.use_constant_memory_space {
+                // We aren't using constant memory, so put the instance in global memory.
+                AddressSpace(1)
```

Comment:
nit: could we make consts somewhere that represent this 0 1 2 3 4 stuff better for easier readability?